traits/ctraits.c

```diff
@@ -1147,16 +1147,26 @@ _has_traits_trait ( has_traits_object * obj, PyObject * args ) {
         }
 
         dict = delegate->obj_dict;
-        if ( ((dict == NULL) ||
-              ((temp_delegate = (has_traits_object *) PyDict_GetItem( dict,
-                                          trait->delegate_name )) == NULL)) &&
-              ((temp_delegate = (has_traits_object *) has_traits_getattro(
-                  delegate, trait->delegate_name )) == NULL) )
-            break;
 
+        temp_delegate = NULL;
+        if (dict != NULL) {
+            temp_delegate = (has_traits_object *) PyDict_GetItem(
```

Comment:
Hmm.  A `NULL` return here indicates that an exception was set.  We should be either clearing the exception or propagating it down the call stack.  As it is, the code just ignores it and continues with the next `if`.